beacon/blockchain/process_proposal.go

```diff
@@ -354,8 +364,66 @@ func (s *Service) verifyStateRoot(
 	return valUpdates, err
 }
 
-// shouldBuildNextPayload returns true if optimistic
-// payload builds are enabled.
+// shouldBuildNextPayload returns true if payload builds are enabled and we are the next block proposer.
 func (s *Service) shouldBuildNextPayload(isNextBlockProposer bool) bool {
 	return isNextBlockProposer && s.localBuilder.Enabled()
 }
+
+// evictLocalPayloadIfNecessary evicts the payload from local builder cache if necessary.
+// We deem it necessary only when we are rejecting the incoming block proposal and it has
+// been built by us.
+func (s *Service) evictLocalPayloadIfNecessary(
+	ctx context.Context,
+	processProposalErr *error,
+	thisNodeAddress []byte,
+	proposerAddress []byte,
+) {
+	// Check conditions for calling this function.
+	// - Function has been called with an initialized error (to hold process proposal error).
+	// - Our node has the local payload builder enabled.
+	if processProposalErr == nil || !s.localBuilder.Enabled() {
+		return
+	}
+
+	// Check the conditions for evicting the local payload.
+	// - There must be a process proposal error (i.e. we are rejecting the proposal).
+	// - The proposer address must be the same as this node's address.
+	if *processProposalErr == nil || !bytes.Equal(thisNodeAddress, proposerAddress) {
+		return
+	}
+
+	// Get the slot and parent block root from the state.
+	st := s.storageBackend.StateFromContext(ctx)
```

Comment:
this is not enough to insulate state from any unnecessary modifications, and you need to add `copiedState := state.Copy(ctx)` and work on the copied state, similarly to what is done in `VerifyIncomingBlock` before `s.preFetchBuildData`